[JIT] Virtualize Function

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34140 [WIP][quantization] Use torchbind for Linear PackedParams
* #34339 [quantization] Make FP16 RNN use new prepack op
* #34098 [JIT] Introduce BuiltinOpFunction and integrate into torchbind
* **#33921 [JIT] Virtualize Function**
* #33900 [JIT] Move stuff out of class_type.cpp

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

Commit:
[JIT] Virtualize Function



Differential Revision: [D20153092](https://our.internmc.facebook.com/intern/diff/D20153092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


Differential Revision: [D20153092](https://our.internmc.facebook.com/intern/diff/D20153092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/api/compilation_unit.h:
```diff
@@ -117,7 +118,7 @@ struct TORCH_API CompilationUnit {
       name = mangle(name);
     }
-    auto fn = torch::make_unique<Function>(
+    auto fn = torch::make_unique<FunctionImpl>(
```
FunctionImpl -> GraphFunction to be more specific since BuiltinFunction is also an implemenation.

jamesr66a replied:
We already have a concept called builtin functions btw: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/builtin_functions.cpp

Should we rename something?

zdevito commented on torch/csrc/jit/api/function.h:
```diff
@@ -17,122 +24,33 @@ TORCH_API void preoptimizeGraph(std::shared_ptr<Graph>& graph);
+
+  virtual size_t num_inputs() const = 0;
+
+  virtual Function& setSchema(c10::FunctionSchema schema) = 0;
```
setSchema/getSchema probably do not need to be virtual. num_inputs and check_single_output as well, and pretty_print_schema.

jamesr66a replied:
Almost all of these functions need to access the IR, which isn't going to be available under an ATen build

zdevito commented on torch/csrc/jit/api/method.h:
```diff
@@ -32,7 +33,7 @@ struct TORCH_API Method {
 
   std::shared_ptr<Graph> graph() const {
-    return function_->graph();
+    return dynamic_cast<FunctionImpl*>(function_)->graph();
```
When builtin-functions are added, this can fail. Also we should avoid dynamic casting and use toGraphFunction/toBuiltinFunction methods on Function.

zdevito commented on torch/csrc/jit/passes/decompose_ops.cpp:
```diff
@@ -108,7 +108,8 @@ bool DecomposeOps(Block* block, script::CompilationUnit& decompose_funcs) {
       WithInsertPoint guard(*it);
       std::shared_ptr<Graph> d_graph =
-          decompose_funcs.get_function("addmm").graph();
+          dynamic_cast<FunctionImpl&>(decompose_funcs.get_function("addmm"))
```
Maybe we should have .graph() but allow it to return a nullptr if it is a built in, it would clean up all these dynamic casts in places where we know there should be a graph representation.

zdevito commented on torch/csrc/jit/passes/subgraph_rewrite.cpp:
```diff
@@ -29,7 +29,7 @@ void SubgraphRewriter::RegisterRewritePattern(
   nodes_to_delete_.clear();
   for (const auto& m : module.get_methods()) {
-    auto g = m.function().graph();
+    auto g = dynamic_cast<FunctionImpl&>(m.function()).graph();
```
I feel like some of these will not always have graphs. For instance, what happens if a module has a builtin method defined on it? We should rewrite these sites to do the right thing, e.g. just not run on non-existing graphs.

zdevito reviewed (commented):
Cool -- let's talk about how to handle .graph() in all the sites that seem to require it.

Commit:
Update on "[JIT] Virtualize Function"


Differential Revision: [D20153092](https://our.internmc.facebook.com/intern/diff/D20153092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20168123](https://our.internmc.facebook.com/intern/diff/D20168123)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20171609](https://our.internmc.facebook.com/intern/diff/D20171609)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20175872](https://our.internmc.facebook.com/intern/diff/D20175872)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/function.h:
```diff
@@ -0,0 +1,62 @@
+
+  virtual GraphExecutor& get_executor() = 0;
+
+  virtual const c10::FunctionSchema& getSchema() const = 0;
```
I still think that at least setSchema and getSchema can be non-virtual. Other things seem good.

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]

Commit:
Update on "[JIT] Virtualize Function"


**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20153092/)!

Differential Revision: [D20177227](https://our.internmc.facebook.com/intern/diff/D20177227)

[ghstack-poisoned]
